homeassistant/components/fritzbox/coordinator.py

```diff
@@ -28,28 +29,48 @@ class FritzboxCoordinatorData:
 class FritzboxDataUpdateCoordinator(DataUpdateCoordinator[FritzboxCoordinatorData]):
     """Fritzbox Smarthome device data update coordinator."""
 
+    config_entry: ConfigEntry
     configuration_url: str
 
-    def __init__(
-        self, hass: HomeAssistant, entry: ConfigEntry, has_templates: bool
-    ) -> None:
+    def __init__(self, hass: HomeAssistant, name: str, has_templates: bool) -> None:
         """Initialize the Fritzbox Smarthome device coordinator."""
-        self.entry = entry
-        self.fritz: Fritzhome = hass.data[DOMAIN][self.entry.entry_id][CONF_CONNECTIONS]
-        self.configuration_url = self.fritz.get_prefixed_host()
-        self.has_templates = has_templates
-        self.new_devices: set[str] = set()
-        self.new_templates: set[str] = set()
-
         super().__init__(
             hass,
             LOGGER,
-            name=entry.entry_id,
+            name=name,
             update_interval=timedelta(seconds=30),
         )
 
+        self.fritz: Fritzhome = hass.data[DOMAIN][self.config_entry.entry_id][
+            CONF_CONNECTIONS
+        ]
+        self.configuration_url = self.fritz.get_prefixed_host()
+        self.has_templates = has_templates
+        self.new_devices: set[str] = set()
+        self.new_templates: set[str] = set()
+
         self.data = FritzboxCoordinatorData({}, {})
 
+    def cleanup_removed_devices(self, avaiable_ains: list[str]) -> None:
+        """Cleanup entity and device registry from removed devices."""
+        entity_reg = er.async_get(self.hass)
+        for entity in er.async_entries_for_config_entry(
+            entity_reg, self.config_entry.entry_id
+        ):
+            if entity.unique_id.split("_")[0] not in avaiable_ains:
+                LOGGER.debug("Removing obsolete entity entry %s", entity.entity_id)
+                entity_reg.async_remove(entity.entity_id)
+
+        device_reg = dr.async_get(self.hass)
+        for device in dr.async_entries_for_config_entry(
+            device_reg, self.config_entry.entry_id
+        ):
+            if not any((DOMAIN, ain) in device.identifiers for ain in avaiable_ains):
```

Comment:
Pre calculate the accepted set at the start of the function so you dont need to redo that.

```
identifiers = {(DOMAIN, ain) for ain in avaiable_ains)}
```
Then you should be able to do:

```
if not set(device.identifiers) & identifiers: 
```

Reply:
efa950af8385ac222d5124b9829647e88f73bcff